File: apps/service-java/src/main/java/com/finflow/model/Transaction.java
package com.finflow.model;

import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Map;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "transactions")
@CompoundIndex(name = "status_date_idx", def = "{'status': 1, 'createdAt': -1}")
@CompoundIndex(name = "category_amount_idx", def = "{'category': 1, 'amount': -1}")
public class Transaction {

    @Id
    private String id;

    @Indexed
    private String userId;

    @Indexed
    private String category;

    private BigDecimal amount;

    private String currency;

    private String description;

    @Indexed
    private TransactionStatus status;

    private TransactionType type;

    private Map<String, Object> metadata;

    private String processedByAgent;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    public enum TransactionStatus {
        PENDING, PROCESSING, COMPLETED, FAILED, CANCELLED
    }

    public enum TransactionType {
        CREDIT, DEBIT, TRANSFER, REFUND
    }
}

File: packages/shared-schemas/src/schemas/transaction.ts
import { z } from 'zod';

export const TransactionStatusEnum = z.enum(['PENDING', 'PROCESSING', 'COMPLETED', 'FAILED', 'CANCELLED']);
export const TransactionTypeEnum = z.enum(['CREDIT', 'DEBIT', 'TRANSFER', 'REFUND']);

export const TransactionSchema = z.object({
  id: z.string(),
  userId: z.string(),
  category: z.string(),
  amount: z.number(),
  currency: z.string().default('USD'),
  description: z.string(),
  status: TransactionStatusEnum,
  type: TransactionTypeEnum,
  metadata: z.record(z.unknown()).optional(),
  processedByAgent: z.string().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export const CreateTransactionSchema = z.object({
  userId: z.string().min(1, 'User ID is required'),
  category: z.string().min(1, 'Category is required'),
  amount: z.number().positive('Amount must be positive'),
  currency: z.string().length(3).default('USD'),
  description: z.string().min(1, 'Description is required'),
  type: TransactionTypeEnum,
  metadata: z.record(z.unknown()).optional(),
});

export type Transaction = z.infer<typeof TransactionSchema>;
export type CreateTransactionInput = z.infer<typeof CreateTransactionSchema>;
